lms/djangoapps/teams/serializers.py

```diff
@@ -0,0 +1,140 @@
+        read_only_fields = ("username",)
+
+
+class CollapsedCourseTeamSerializer(serializers.HyperlinkedModelSerializer):
+    """
+    Serializes CourseTeams in a collapsed format, with just an id and url.
+    """
+    id = serializers.CharField(source='team_id', read_only=True)  # pylint: disable=invalid-name
+    url = serializers.HyperlinkedIdentityField(view_name='teams_detail', lookup_field='team_id')
+
+    class Meta(object):
+        """
+        Defines meta information for the ModelSerializer.
+        """
+        model = CourseTeam
+        fields = ("id", "url")
+
+
+class UserMembershipSerializer(serializers.ModelSerializer):
+    """
+    Serializes CourseTeamMemberships with only information about the user and date_joined, for listing team members.
+    """
+    user = ExpandableField(
+        collapsed_serializer=CollapsedUserSerializer(),
+        expanded_serializer=UserSerializer(),
+    )
+
+    class Meta(object):
+        """
+        Defines meta information for the ModelSerializer.
+        """
+        model = CourseTeamMembership
+        fields = ("user", "date_joined")
+        read_only_fields = ("date_joined",)
+
+
+class MembershipSerializer(serializers.ModelSerializer):
+    """
+    Serializes CourseTeamMemberships with information about both teams and users.
+    """
+    user = CollapsedUserSerializer(read_only=True)
+    team = CollapsedCourseTeamSerializer(read_only=True)
+
+    class Meta(object):
+        """
+        Defines meta information for the ModelSerializer.
+        """
+        model = CourseTeamMembership
+        fields = ("user", "team", "date_joined")
+        read_only_fields = ("date_joined",)
+
+
+class CourseTeamSerializer(serializers.ModelSerializer):
+    """
+    Serializes a CourseTeam with membership information.
+    """
+    id = serializers.CharField(source='team_id', read_only=True)  # pylint: disable=invalid-name
+    membership = UserMembershipSerializer(many=True, read_only=True)
+
+    class Meta(object):
+        """
+        Defines meta information for the ModelSerializer.
+        """
+        model = CourseTeam
+        fields = (
+            "id",
+            "name",
+            "is_active",
+            "course_id",
+            "topic_id",
+            "date_created",
+            "description",
+            "country",
+            "language",
+            "membership",
+        )
+        read_only_fields = ("course_id", "date_created")
+
+
+class CourseTeamCreationSerializer(serializers.ModelSerializer):
+    """
+    Deserializes a CourseTeam for creation.
+    """
+
+    class Meta(object):
+        """
+        Defines meta information for the ModelSerializer.
+        """
+        model = CourseTeam
+        fields = (
+            "name",
+            "course_id",
+            "description",
+            "topic_id",
+            "country",
+            "language",
+        )
+
+    def restore_object(self, attrs, instance=None):
```

Comment:
I need to manually restore the object because it constructs its `team_id` on the fly through `CourseTeam.create()`. The default ModelSerializer restore_object() will simply try to restore each field on the object from the given information. Granted, my understanding of serializes is still a bit fuzzy, and there may be a better way to do this.